test/test_type_promotion.py

```diff
@@ -343,19 +343,69 @@ def test_create_bool_tensors(self, device):
+        a_single_tensor = torch.tensor(1, device=device, dtype=dtypes[0])
+        a_scalar = a_single_tensor.item()
+        b_tensor = torch.tensor((1, 0), device=device, dtype=dtypes[1])
+        b_single_tensor = torch.tensor(1, device=device, dtype=dtypes[1])
+        b_scalar = b_single_tensor.item()
+        for a, b in ((a_tensor, b_tensor), (a_single_tensor, b_single_tensor), (a_scalar, b_scalar)):
+            dtype_a = _get_dtype(a)
+            dtype_b = _get_dtype(b)
+            try:
+                result = a + b
+            except RuntimeError:
+                with self.assertRaises(RuntimeError):
+                    torch.promote_types(dtype_a, dtype_b)
+                with self.assertRaises(RuntimeError):
+                    torch.result_type(a, b)
+            else:
+                dtype_res = _get_dtype(result)
+                if a is a_scalar and b is b_scalar and dtype_a == torch.bool and dtype_b == torch.bool:
+                    # special case: in Python, True + True is an integer
+                    self.assertEqual(dtype_res, torch.int64, f"a == {a}, b == {b}")
+                else:
+                    self.assertEqual(dtype_res, torch.result_type(a, b), f"a == {a}, b == {b}")
+                if a is a_scalar and b is b_scalar:  # Python internal type determination is good enough in this case
+                    continue
+                self.assertEqual(dtype_res, torch.promote_types(dtype_a, dtype_b), f"a == {a}, b == {b}")
+
     @float_double_default_dtype
-    def test_result_type(self, device):
-        self.assertEqual(torch.result_type(torch.tensor(1, dtype=torch.int, device=device), 1), torch.int)
-        self.assertEqual(torch.result_type(1, torch.tensor(1, dtype=torch.int, device=device)), torch.int)
-        self.assertEqual(torch.result_type(1, 1.), torch.get_default_dtype())
-        self.assertEqual(torch.result_type(torch.tensor(1, device=device), 1.), torch.get_default_dtype())
-        self.assertEqual(torch.result_type(torch.tensor(1, dtype=torch.long, device=device),
-                         torch.tensor([1, 1], dtype=torch.int, device=device)),
-                         torch.int)
-        self.assertEqual(torch.result_type(torch.tensor([1., 1.], dtype=torch.float, device=device), 1.), torch.float)
-        self.assertEqual(torch.result_type(torch.tensor(1., dtype=torch.float, device=device),
-                         torch.tensor(1, dtype=torch.double, device=device)),
-                         torch.double)
+    def test_result_type_tensor_vs_scalar(self, device):
+        "Spot check some result type for tensor against scalar (including single-element tensor)."
+
+        def _test_spot(a, b, res_dtype):
+            self.assertEqual(torch.result_type(a, b), res_dtype)
+            self.assertEqual(torch.result_type(b, a), res_dtype)
+
+        _test_spot(torch.tensor(1, dtype=torch.int, device=device), 1, torch.int)
```

Comment:
Add two 1D+ tensor cases:
- half x long -> half
- float x double -> double

And one 0D tensor case:
- complex64 x complex128 -> complex128

Reply:
I added a few spot checks. Since 1D+ x 1D+ are already tested in the test above, I believe you actually meant 1D+ x 0D.